### PR TITLE
[tosa] Add maxpool2d and adaptive_avgpool2d support

### PR DESCRIPTION
**Pull request by sjarus**
Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

### CODE REVIEW

**Comment by sjarus** [comment]
Testing: 

Both e2e tests pass with shapes made static: 

****** Unexpectedly Passed tests - 2 tests
    XPASS - "AdaptiveAvgPool2dModule_basic"
    XPASS - "MaxPool2dModule_basic"

A very large unit test suite also passes all permutations of inputs given: 

Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out1x1
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out1x8
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out1x16
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out8x1
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out8x8
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out8x16
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out16x1
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out16x8
Results PASS test_adaptive_avgpool2d_1x1x16x16_f32_out16x16
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out1x1
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out1x16
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out1x32
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out16x1
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out16x16
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out16x32
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out32x1
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out32x16
Results PASS test_adaptive_avgpool2d_5x2x32x32_f32_out32x32
Results PASS test_maxpool2d_3x3_1x1x16x16_f32_st11_pad00
Results PASS test_maxpool2d_3x3_1x1x16x16_f32_st12_pad00
Results PASS test_maxpool2d_3x3_1x1x16x16_f32_st21_pad00
Results PASS test_maxpool2d_3x3_1x1x16x16_f32_st22_pad00
Results PASS test_maxpool2d_3x3_5x2x32x32_f32_st11_pad00
Results PASS test_maxpool2d_3x3_5x2x32x32_f32_st12_pad00
Results PASS test_maxpool2d_3x3_5x2x32x32_f32_st21_pad00
Results PASS test_maxpool2d_3x3_5x2x32x32_f32_st22_pad00
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st11_pad00
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st11_pad01
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st11_pad10
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st11_pad11
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st12_pad00
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st12_pad01
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st12_pad10
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st12_pad11
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st21_pad00
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st21_pad01
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st21_pad10
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st21_pad11
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st22_pad00
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st22_pad01
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st22_pad10
Results PASS test_maxpool2d_5x5_1x1x16x16_f32_st22_pad11
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st11_pad00
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st11_pad01
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st11_pad10
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st11_pad11
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st12_pad00
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st12_pad01
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st12_pad10
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st12_pad11
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st21_pad00
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st21_pad01
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st21_pad10
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st21_pad11
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st22_pad00
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st22_pad01
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st22_pad10
Results PASS test_maxpool2d_5x5_5x2x32x32_f32_st22_pad11

**Comment by sjarus** [comment]
With a few changes to RefineTypes to be released separately, this also passes static shaped ResNet18 - the first full network on TorchToTosa: 

$ ./tools/torchscript_e2e_test.sh --config=tosa --filter=ResNet18StaticModule_basic -v
PASS - "ResNet18StaticModule_basic"

Summary:
    Passed: 1

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -1987,6 +1987,278 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
+    else
+      return (
+          (inputDim + padBefore + padAfter - dilation * (kernelDim - 1) - 1) /
+              stride +
+          1);
+  }
+
+  // Apply the transposeDims vector on input to generate a transposed form.
+  Value transposeTensor(AtenOpT op, ConversionPatternRewriter &rewriter,
+                        Value input, SmallVector<int32_t> transposeDims) const {
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2022]: it's recommended to use ArrayRef if the list is not modified. refer to https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -1987,6 +1987,278 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
+
+    SmallVector<int64_t> transposedInputShape;
+    for (auto &dim : transposeDims)
+      transposedInputShape.push_back(inputShape[dim]);
+    auto transposedInputType =
+        RankedTensorType::get(transposedInputShape, inputElemTy);
+    return rewriter
+        .create<tosa::TransposeOp>(
+            op->getLoc(),
+            OpConversionPattern<AtenOpT>::getTypeConverter()->convertType(
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2041]: `transposedInputType ` is built in tensor type already so the type conversion is not needed.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -1987,6 +1987,278 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
+    SmallVector<int32_t> nhwcToNchw4DTransposeDims({0, 3, 1, 2});
+    SmallVector<int32_t> hwcToChw3DTransposeDims({2, 0, 1});
+
+    return transposeTensor(op, rewriter, input,
+                           inputRank == 3 ? hwcToChw3DTransposeDims
+                                          : nhwcToNchw4DTransposeDims);
+  }
+
+  LogicalResult matchAndRewrite(AtenOpT op, OpAdaptor adaptor,
+                                ConversionPatternRewriter &rewriter) const {
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2076]: need to mark `override`

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -1987,6 +1987,278 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
+    auto inputTy = inputXchw.getType().template cast<RankedTensorType>();
+    if (!inputTy)
+      return op.emitError("Adaptive avgpool requires ranked tensor input");
+
+    auto inputShape = inputTy.getShape();
+    auto inputRank = inputTy.getRank();
+    auto inputElemTy = inputTy.getElementType();
+
+    // Rank sanity check.
+    if (!(inputTy.getRank() == 4) && !(inputRank == 3))
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2126]: nit: inputTy.getRank != 4 and inputRank != 3

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 5]
```diff
@@ -1987,6 +1987,278 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2205]: nit: inputTy.getRank() != 4 and inputRank != 3

**Review by cathyzhyi — CHANGES_REQUESTED** [review]
Looks good other than a few comments.

**Comment by sjarus** [comment]
@cathyzhyi I've tried to duplicate the failure locally but am unable to do so - it passes e2e locally. Is there something obvious I'm missing here ?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 6]
```diff
@@ -1987,6 +1987,277 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
+  }
+};
+
+template <typename AtenOpT, typename TosaOpT>
+class ConvertAtenAdaptivePoolingOp
+    : public ConvertAtenPoolingBaseOp<AtenOpT, TosaOpT> {
+public:
+  using ConvertAtenPoolingBaseOp<AtenOpT, TosaOpT>::ConvertAtenPoolingBaseOp;
+  using OpAdaptor = typename AtenOpT::Adaptor;
+  LogicalResult processInputs(AtenOpT op, OpAdaptor adaptor,
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2111]: need `override` here as well

**Comment by cathyzhyi** [comment]
> @cathyzhyi I've tried to duplicate the failure locally but am unable to do so - it passes e2e locally. Is there something obvious I'm missing here ?

hmmm, it's not obvious to me. Let me try on my machine.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 7]
```diff
@@ -1987,6 +1987,277 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
...
+        .getResult();
+  }
+
+  Value transposePoolingInputToHwc(AtenOpT op,
+                                   ConversionPatternRewriter &rewriter,
+                                   Value input) const {
+    auto inputRank =
+        input.getType().template cast<RankedTensorType>().getRank();
+
+    ArrayRef<int32_t> nchwToNhwc4DTransposeDims({0, 2, 3, 1});
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2051]: oh, you still need to use SmallVector here. ArrayRef won't keep the integer aline.

**Comment by sjarus** [comment]
Passes now, thanks for that ArrayRef catch. Also found another missing override I added - the build has no warnings now.

**Review by cathyzhyi — APPROVED** [review]

**Pull request merged**